packages/next/next-server/lib/mitt.ts

```diff
@@ -15,31 +15,21 @@ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLI
 // See the LICENSE at the top of the file
 
 type Handler = (...evts: any[]) => void
-
-export type MittEmitter = {
-  on(type: string, handler: Handler): void
-  off(type: string, handler: Handler): void
-  emit(type: string, ...evts: any[]): void
-}
-
-export default function mitt(): MittEmitter {
-  const all: { [s: string]: Handler[] } = Object.create(null)
-
+export default function mitt<T extends string, H extends Handler = Handler>() {
```

Comment:
The handler changes per event type, so you'll likely want something like this: https://rjzaworski.com/2019/10/event-emitters-in-typescript

However in our case we have a handler for `routeChangeError` that has 2 parameters instead of one.

Reply:
Hey @timneutkens  

I did an event emitter for the router events, refer to https://codesandbox.io/s/nextjs-events-2in1n?file=/src/index.ts:785-1124

But I'm having some doubts in order to implement it with the current ``mitt`` 

In order to type the handlers, we should pass an event map as a generic. 
```
interface Mitt<T extends RouterEventMap> {
  on<K extends RouterEventKey<RouterEventMap>>(
    type: K,
    handler: RouterEventHandler<K>
  ): void;
  off<K extends RouterEventKey<RouterEventMap>>(
    type: K,
    handler: RouterEventHandler<K>
  ): void;
  emit<K extends RouterEventKey<RouterEventMap>>(type: K, ...evts: T[K]): void;
}
```

But the ``mitt`` function it's being used in two places: 

Inside of the ``router.ts``: 
```
  static events = mitt<RouterEvent, (url: string) => void>()
```

Inside of the class ``ServerRouter``: 

```
  static events = mitt<string>()
```

The generic being passed to the ``mitt`` function above it's just a string, so there are two different cases. 

How can we pass the ``RouterEventMap`` as a generic for the router emitter and a ``string`` as a generic for the ``ServerRouter`` events emitter?

Reply:
The one in `ServerRouter` could just be typed `any` I think, `ServerRouter` never uses the event emitter and it's purely there for backwards compatiblity 👍

Reply:
@timneutkens Just added the router events map, what do you think?